WordPress/gutenberg
# Layout: Add layout aware margin styles output for blocks in global styles

## What?


The idea in this PR is to add additional output for margin styles in themes that opt-in to blockGap, to introduce layout aware margin rules that are particular to the layout types. Since the existing flow and constrained layouts borrow their margin/spacing logic from the idea of[ the Stack layout as described in Every Layout book](https://every-layout.dev/layouts/stack/), the approach in this PR is inspired by the idea of exceptions, by outputting additional margin rules. Where the approach in this PR differs is that it uses direct values as output, rather than CSS variables.

So, in a block theme that opts-in to block gap, if a user or theme sets margin styles on the Group block, then this PR will output additional rules. Example output:

```css
/* Standard margin output is still output so that blocks that appear outside of layout blocks still receive margin rules */
.wp-block-group {
  margin-top: 25px;
  margin-bottom: 50px;
}
/* Layout aware margin output, to override margins for the direct child of the layout type, if it matches the block's selector */
.is-layout-flow > .wp-block-group {
  margin-top:25px;
  margin-bottom:50px;
}'
```

## Why?


As described in #43404, the existing layout rules currently override margin rules set at the block level in global styles. A previous attempt looked at trying to reduce the specificity of layout styles (https://github.com/WordPress/gutenberg/pull/45927), however this approach did not result in the desired outcome. The specificity of layout styles was reduced, however it then resulted in the layout and margin rules having the same level of specificity (010), which meant that if the layout styles were output after the margin styles, the layout styles still won out.

So, the argument in this PR is that it would be better to be explicit about having block-level margin rules that intentionally have a higher specificity than layout rules, however lower specificity than layout rules set at the individual block level. The result is:

* `.is-layout-flow * + *` (010) — base layout styles
* `.is-layout-flow > .wp-block-group` (020) — layout aware margin styles output

## How?


* Add a `marginSelector` property to the layout definitions in `theme.json`.
* In both the PHP and JS global styles output, check if block gap is supported by the theme and if margin is supported by the current block — if so, then output layout aware margin rules as above, using the `marginSelector` to join the layout classname with the block's classname.
* Update tests to match.

## Testing Instructions





1. In the site editor, set some margins for blocks that support it (e.g. Cover or Heading blocks)
2. Add blocks within group blocks and ensure margins are being output correctly
3. Save and view the site frontend — margin rules should win out over base layout rules
4. Back in the site editor (or post editor) set a block spacing value on an individual container block (e.g. Group) with a Cover block inside it — the block spacing value at the individual container block should override the margin rules for the Cover block from global styles
5. Ensure all these styles are working correctly in the post editor
6. (Regression testing) Double-check that the changes here do not unintentionally spill any gap or margin rules into themes that should not have theme (e.g. if block gap is set to `null` in `theme.json` or if running a Classic theme)

A question — do we also need to output these extra rules for the `flex` layout? I don't _think_ so, since it uses `gap` rather than vertical margins, but it might be worth double checking.

### Testing Instructions for Keyboard


## Screenshots or screencast 

| Before (note that the Cover block's margin rules do not appear to apply) | After (note that the Cover block's margin rules do apply |
| --- | --- |
| <img width="685" alt="image" src="https://user-images.githubusercontent.com/14988353/214486062-ffb4f63f-702d-4a8d-8a6b-03520572be09.png"> | <img width="694" alt="image" src="https://user-images.githubusercontent.com/14988353/214486255-7e481017-f17e-4614-a17c-cd0b8856c354.png"> |
